safe_prime_generator.go

```diff
@@ -0,0 +1,301 @@
+	case err := <-errChan:
+		mutex.Lock()
+		cancel()
+		mutex.Unlock()
+		p, q, err = nil, nil, err
+	case <-ctx.Done():
+		p, q, err = nil, nil, fmt.Errorf("generator timed out after %v", timeout)
+	}
+
+	waitGroup.Wait()
+	return
+}
+
+type safePrime struct {
+	p *big.Int // p = 2q + 1
+	q *big.Int
+}
+
+// Starts a Goroutine searching for a safe prime of the specified `pBitLen`.
+// If succeeds, writes prime `p` and prime `q` such that `p = 2q+1` to the
+// `primeChan`. Prime `p` has a bit length equal to `pBitLen` and prime `q` has
+// a bit length equal to `pBitLen-1`.
+//
+// The algorithm is as follows:
+// 1. Generate a random odd number `q` of length `pBitLen-1` with two the most
+//    significant bits set to `1`.
+// 2. Execute preliminary primality test on `q` checking whether it is coprime
+//    to all the elements of `smallPrimes`. It allows to eliminate trivial
+//    cases quickly, when `q` is obviously no prime, without running an
+//    expensive final primality tests.
+//    If `q` is coprime to all of the `smallPrimes`, then go to the point 3.
+//    If not, add `2` and try again. Do it at most 10 times.
+// 3. Check the potentially prime `q`, whether `q = 1 (mod 3)`. This will
+//    happen for 50% of cases.
+//    If it is, then `p = 2q+1` will be a multiple of 3, so it will be obviously
+//    not a prime number. In this case, add `2` and try again. Do it at most 10
+//    times. If `q != 1 (mod 3)`, go to the point 4.
+// 4. Now we know `q` is potentially prime and `p = 2q+1` is not a multiple of
+//    3. We execute a preliminary primality test on `p`, checking whether
+//    it is coprime to all the elements of `smallPrimes` just like we did for
+//    `q` in point 2. If `p` is not coprime to at least one element of the
+//    `smallPrimes`, then go back to point 1.
+//    If `p` is coprime to all the elements of `smallPrimes`, go to point 5.
+// 5. At this point, we know `q` is potentially prime, and `p=q+1` is also
+//    potentially prime. We need to execute a final primality test for `q`.
+//    We apply Miller-Rabin and Baillie-PSW tests. If they succeed, it means
+//    that `q` is prime with a very high probability. Knowing `q` is prime,
+//    we use Pocklington's criterion to prove the primality of `p=2q+1`, that
+//    is, we execute Fermat primality test to base 2 checking whether
+//    `2^{p-1} = 1 (mod p)`. It's significantly faster than running full
+//    Miller-Rabin and Baillie-PSW for `p`.
+//    If `q` and `p` are found to be prime, return them as a result. If not, go
+//    back to the point 1.
+func runGenPrimeRoutine(
+	ctx context.Context,
+	primeChan chan safePrime,
+	errChan chan error,
+	mutex *sync.Mutex,
+	waitGroup *sync.WaitGroup,
+	rand io.Reader,
+	pBitLen int,
+) {
+	qBitLen := pBitLen - 1
+	b := uint(qBitLen % 8)
+	if b == 0 {
+		b = 8
+	}
+
+	bytes := make([]byte, (qBitLen+7)/8)
+	p := new(big.Int)
+	q := new(big.Int)
+
+	bigMod := new(big.Int)
+
+	go func() {
```

Comment:
You can put `defer waitGroup.Done()` here and remove the other two references. That way we know that we'll call `Done` regardless of what path we trigger!